nemo/collections/nlp/models/multi_label_intent_slot_classification/multi_label_intent_slot_classification_model.py

```diff
@@ -0,0 +1,406 @@
+            shuffle=cfg.shuffle,
+            num_workers=cfg.num_workers,
+            pin_memory=cfg.pin_memory,
+            drop_last=cfg.drop_last,
+            collate_fn=dataset.collate_fn,
+        )
+
+    def prediction_probabilities(self, queries: List[str], test_ds) -> List[List[float]]:
+        """
+        Get prediction probabilities for the queries (intent and slots)
+        Args:
+            queries: text sequences
+            test_ds: Dataset configuration section.
+        Returns:
+            predicted_intents, predicted_slots: model intent and slot predictions
+        """
+
+        probabilities = []
+
+        mode = self.training
+        try:
+            device = "cuda" if torch.cuda.is_available() else "cpu"
+
+            # Initialize tokenizer.
+            # if not hasattr(self, "tokenizer"):
+            #    self._setup_tokenizer(self.cfg.tokenizer)
+            # Initialize modules.
+            # self._reconfigure_classifier()
+
+            # Switch model to evaluation mode
+            self.eval()
+            self.to(device)
+
+            # Dataset.
+            infer_datalayer = self._setup_infer_dataloader(queries, test_ds)
+
+            for batch in infer_datalayer:
+                input_ids, input_type_ids, input_mask, loss_mask, subtokens_mask = batch
+
+                intent_logits, slot_logits = self.forward(
+                    input_ids=input_ids.to(device),
+                    token_type_ids=input_type_ids.to(device),
+                    attention_mask=input_mask.to(device),
+                )
+
+                # predict intents and slots for these examples
+                # intents
+                probabilities.append(torch.sigmoid(intent_logits))
+
+            probabilities = torch.cat(probabilities)
+
+        finally:
+            # set mode back to its original value
+            self.train(mode=mode)
+
+        return probabilities
+
+    def optimize_threshold(self, test_ds):
+        input_file = f"{self.data_dir}/dev.tsv"
+
+        with open(input_file, "r") as f:
+            input_lines = f.readlines()[1:]  # Skipping headers at index 0
+
+        dataset = list(input_lines)
+
+        metrics_labels, sentences = [], []
+
+        for input_line in dataset:
+            sentence = input_line.strip().split("\t")[0]
+            sentences.append(sentence)
+            parts = input_line.strip().split("\t")[1:][0]
+            parts = list(map(int, parts.split(",")))
+            parts = [1 if label in parts else 0 for label in range(len(self.cfg.data_desc.intent_labels))]
+            metrics_labels.append(parts)
+
+        # Retrieve class probabilities for each sentence
+        intent_probabilities = self.prediction_probabilities(sentences, test_ds)
+
+        metrics_dict = {}
+        # Find optimal logits rounding threshold for intents
+        for i in np.arange(0.5, 0.96, 0.01):
+            predictions = tensor2list((intent_probabilities >= i).int())
```

Comment:
This line will also need to be modified if you convert the torch tensor to a numpy array in `prediction_probabilities`. You can use numpy `.tolist()`